// web3-react/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var provider_1 = __importDefault(require("./provider"));
exports.default = provider_1.default;
var provider_2 = require("./provider");
exports.useWeb3Context = provider_2.useWeb3Context;
exports.Web3Consumer = provider_2.Web3Consumer;
exports.withWeb3 = provider_2.withWeb3;
var manager_1 = require("./manager");
exports.ManagerErrorCodes = manager_1.ManagerErrorCodes;
var subproviders = __importStar(require("@0x/subproviders"));
exports.subproviders = subproviders;
var Connectors = __importStar(require("./connectors"));
exports.Connectors = Connectors;
